src/app/components/import-annotations/add-annotations/add-annotations.component.ts

```diff
@@ -0,0 +1,364 @@
+
+  /**
+   * A boolean value indicating if the currently buffered audio event import can
+   * be successfully committed
+   */
+  protected valid: boolean = false;
+  protected uploading: boolean = false;
+  protected importFiles: File[] = [];
+  protected additionalTagIds: Id[] = [];
+
+  /**
+   * Errors that apply to the entire file that we cannot display in the events
+   * table.
+   * E.g. duplicate file uploads, names, etc...
+   */
+  protected importErrors: BawErrorData[] = [];
+  // we use a BehaviorSubject for the audio event import files because users can upload
+  // multiple files through the file input and new subscribers should get the most recent value
+  protected importFiles$ = new BehaviorSubject<AudioEventImportFile[]>([]);
+
+  public get hasUnsavedChanges(): boolean {
+    return this.hasFiles || this.hasAdditionalTags;
+  }
+
+  // if the "Import Annotations" button is disabled, we want to provide some
+  // feedback to the user outlining why they cannot submit the form using a
+  // tooltip
+  protected get uploadTooltip(): string | null {
+    // by returning null, the tooltip will not be displayed
+    if (this.canCommitUploads()) {
+      return null;
+    }
+
+    if (!this.hasFiles) {
+      return "Please select a file to upload";
+    } else if (!this.valid || !this.form.valid) {
+      return "Please fix all errors before submitting";
+    } else if (this.uploading) {
+      return "Please wait for the current upload to complete";
+    }
+
+    // we should never hit this condition, but if we do, I return null so that
+    // we don't end up in an unknown state
+    return null;
+  }
+
+  private get hasFiles(): boolean {
+    return this.importFiles.length > 0;
+  }
+
+  private get hasAdditionalTags(): boolean {
+    return this.additionalTagIds.length > 0;
+  }
+
+  public ngOnInit(): void {
+    const models = retrieveResolvers(this.route.snapshot.data as IPageInfo);
+
+    if (!hasResolvedSuccessfully(models)) {
+      this.failure = true;
+      return;
+    }
+
+    this.audioEventImport = models[audioEventImportKey] as AudioEventImport;
+  }
+
+  protected getEventModels = (): Observable<ImportedAudioEvent[]> => {
+    return this.importFiles$.pipe(
+      mergeMap((files: AudioEventImportFile[]) => {
+        return files.map((file) => file.importedEvents);
+      })
+    );
+  };
+
+  // callback used by the typeahead input to search for associated tags
+  protected searchTagsTypeaheadCallback = (
+    text: string,
+    activeItems: Tag[]
+  ): Observable<Tag[]> =>
+    this.tagsApi.filter({
+      filter: filterAnd(
+        contains<Tag, "text">("text", text),
+        notIn<Tag>("text", activeItems)
+      ),
+    });
+
+  // since the typeahead input returns an array of models, but the api wants the associated tags as an array of id's
+  // we use this helper function to convert the array of models to an array of id's that can be sent in the api request
```

Comment:
Perhaps the filter should accept an array of tags or ids? We use that convention elsewhere

```suggestion
  // typeahead returns models, we need ids for filter
```